src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,164 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> newNode = new Node(value, tail, null);
+        if (isEmpty()) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) { //add on tail
+            add(value);
+            return;
+        }
+
+        if (index == 0) { //add on head
+            Node<T> newNode = new Node(value, null, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+
+        if (isIndexInvalid(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        Node<T> nodeAfter = findNodeByIndex(index);
 
+        Node<T> newNode = new Node(value, nodeAfter.prev, nodeAfter);
+        nodeAfter.prev.next = newNode;
+        nodeAfter.prev = newNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (isIndexInvalid(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+        return findNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (isIndexInvalid(index)) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        Node<T> settedNode = findNodeByIndex(index);
+
+        T oldValue = settedNode.value;
+        settedNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
how about this comment?
we can create method unlinkNode()/deleteNode()
```
    private T unlink(Node<T> currentNode){
        Node<T> prev = currentNode.prev;
        Node<T> next = currentNode.next;
        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
        }
        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
        }
        size--;
        return currentNode.value;
    }
```
and use it in the method remove by index:
```
        Node<T> removedNode = findNodeByIndex(index);
        return unlink(removedNode);
```
and method remove by value:
```
...
            if (currentValue == null ? currentValue == object : currentValue.equals(object)) {
                unlink(current);
                return true;
            }
...
```